### ISSUE TITLE
Thread Pool: no nice way to do a cleanup

### ISSUE THREAD

**Issue by mijay** [andywer/threads.js#48]
I use the thread pool to process bunch of tasks, but I need to do some cleanup in every worker thread before shutting it down (at least in case when no unexpected errors occurred). As `child_process.kill()` is used under the hood for `pool.killAll()`, I cannot use `process.on("exit"/"beforeExit")` to schedule such cleanup. Currently I could not find a nice way to do so.

I could come up with two a bit hacky solutions:
```javascript
const pool = new Pool(SIZE).run(__dirname + "/runner");
Promise.all(
  workItems.map(work => pool.send({ work }).promise()
)
  .then(result => {
      return Array.from(new Array(SIZE)).map(() => pool.send({ exit: true }).promise()); // 1
      return pool.threads.map(thread => thread.send({ exit: true }).promise()); // 2
  })
  .then(() => pool.killAll())
```
- but the solution 1 relies on the way how `Pool` schedules the work to threads, which is not documented and might change in the future;
- and solution 2 relies on not public api of `Pool` + breaks in cases when `workItems.length < SIZE` as it produces `No thread logic initialized.` errors.

Can you maybe suggest a better way of solving this? Or even supporting a graceful shutdown on the library level? Other option might be to have some api like `sendAll(message)` in `Pool`.

**Comment by andywer** [comment]
Yeah, the pool shutdown is a bit tricky, since usually the pool is not terminated right away after doing an initial chunk of work.

That `kill` vs `on('exit')` situation is interesting. Maybe it might be a good idea to change the `.kill()` behavior, like performing a `child.exit()` and setting some timeout to only call `child.kill()` if it didn't already shut down during this timeout?

**Comment by mijay** [comment]
I would argue that `sendAll` would be both simpler to implement and more logical api to have in `Pool`. Plus that can be easily implemented cross-platform. Of course, then everyone who want to support cleanup scenario like I've described, will have to do it themselves, but that sounded more like a FAQ topic.

**Comment by andywer** [comment]
Fair enough, that is a good argument.

Just out of curiosity... What kind of clean-up do you do?

**Comment by andywer** [comment]
If you need the `sendAll` urgently I'd say open a PR and get credited in the changelog. I might implement it as well, but it's not a high priority to me.

Anyway I can publish v0.7.3 containing the unhandled promise rejection handler today.

**Comment by mijay** [comment]
I ended up using hack 1 and it works for now, just not beautiful enough - so no urgency.

I am using a pool of runners to run bunch of selenium tests in parallel, so I need to shutdown the driver used by each thread properly when there are no more tests left.

**Comment by ZenModeRy** [comment]
Just adding to this… I'm having trouble releasing my database connection from within a worker when the worker fails prematurely. That's my primary use case for cleanup. That and global.gc() when running Node with the `--expose-gc` flag.

**Comment by andywer** [comment]
@ZenModeRy Thanks for sharing your thoughts. To make `if (typeof job.cleanup === 'function')` possible, we would need to extend the API of `Job`, so it's not exactly a "done in ten minutes" change.

But a proper way to clean-up after a job would be nice. I will put it on the "version 1.0 wishlist" #54 😉

**Comment by andywer** [comment]
I think this should be fixed in v1.0. Closing. Feel free to re-open if it still happens in 1.0.
